gitlab/resource_gitlab_project_approval_rule.go

```diff
@@ -0,0 +1,254 @@
+	}
+	d.Set("project", projectID)
+	d.SetPartial("project")
+
+	rule, err := getApprovalRuleByID(meta.(*gitlab.Client), d.Id())
+	if err != nil {
+		return err
+	}
+
+	d.Set("name", rule.Name)
+	d.SetPartial("name")
+
+	d.Set("approvals_required", rule.ApprovalsRequired)
+	d.SetPartial("approvals_required")
+
+	if err := d.Set("group_ids", flattenApprovalRuleGroupIDs(rule.Groups)); err != nil {
+		return err
+	}
+	d.SetPartial("group_ids")
+
+	if err := d.Set("eligible_approvers", flattenApprovalRuleUserIDs(rule.EligibleApprovers)); err != nil {
+		return err
+	}
+	d.SetPartial("eligible_approvers")
+
+	if err := d.Set("user_ids", flattenApprovalRuleUserIDs(rule.Users)); err != nil {
+		return err
+	}
+	d.SetPartial("user_ids")
+
+	d.Partial(false)
+
+	return nil
+}
+
+func resourceGitlabProjectApprovalRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	projectID, ruleID, err := splitApprovalRuleID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	ruleIDInt, err := strconv.Atoi(ruleID)
+	if err != nil {
+		return err
+	}
+
+	options := gitlab.UpdateProjectLevelRuleOptions{
+		Name:              gitlab.String(d.Get("name").(string)),
+		ApprovalsRequired: gitlab.Int(d.Get("approvals_required").(int)),
+		UserIDs:           expandApproverIds(d.GetOk("user_ids")),
+		GroupIDs:          expandApproverIds(d.GetOk("group_ids")),
+	}
+
+	log.Printf("[DEBUG] Project %s update gitlab project-level approval rule %s", projectID, *options.Name)
+
+	client := meta.(*gitlab.Client)
+
+	_, _, err = client.Projects.UpdateProjectApprovalRule(projectID, ruleIDInt, &options)
+	if err != nil {
+		return err
+	}
+
+	return resourceGitlabProjectApprovalRuleRead(d, meta)
+}
+
+func resourceGitlabProjectApprovalRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	project, ruleID, err := splitApprovalRuleID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	ruleIDInt, err := strconv.Atoi(ruleID)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Project %s delete gitlab project-level approval rule %d", project, ruleIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	_, err = client.Projects.DeleteProjectApprovalRule(project, ruleIDInt)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getApprovalRuleByID checks the list of rules and finds the one that matches our rule ID.
+func getApprovalRuleByID(client *gitlab.Client, id string) (*gitlab.ProjectApprovalRule, error) {
+	projectID, ruleID, err := splitApprovalRuleID(id)
+	if err != nil {
+		return nil, err
+	}
+
+	ruleIDInt, err := strconv.Atoi(ruleID)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[DEBUG] read approval rules for project %s", projectID)
+
+	rules, _, err := client.Projects.GetProjectApprovalRules(projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range rules {
+		if r.ID == ruleIDInt {
+			log.Printf("[DEBUG] found project-level rule %+v", r)
+			return r, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find GitLab approval rule %d", ruleIDInt)
```

Comment:
I've updated to set the ID to "" here: https://github.com/gitlabhq/terraform-provider-gitlab/pull/250/files#diff-3d7670821c4b22becad8070158180cb5b64f390c1c0dac4d96b5b74d6c3f84fcR88

Reply:
With the change, you're setting the ID to "" on *any* error. Can you make it so it only sets to "" when the rule is not found?

Reply:
Per the linked PR, this method would:
```
d.SetId("") 
return nil, nil
```

Which means the [next line](https://github.com/gitlabhq/terraform-provider-gitlab/pull/250/files#diff-3d7670821c4b22becad8070158180cb5b64f390c1c0dac4d96b5b74d6c3f84fcR92) of 
```
d.Set("name", rule.Name)
```
Would error since rule is `nil`

Is that correct? Am I misunderstanding something? 

Thanks!

Reply:
The problem is this code

```go
	rule, err := getApprovalRuleByID(meta.(*gitlab.Client), d.Id())
	if err != nil {
		d.SetId("")
		return nil
	}
```

does not check the error cause before setting the ID to "", so it could be some other error like an HTTP 500 from the GitLab API.

There are a couple ways to work around this. You could add a third return argument to `getApprovalRuleByID`, type `bool`, to signal if the rule was not found, and check that bool inside the error check where you call it. You could also create the error as a global variable and check it where you call `getApprovalRuleByID`, like:

```go
var errApprovalRuleNotFound = errors.New("approval rule not found")

...

	rule, err := getApprovalRuleByID(meta.(*gitlab.Client), d.Id())
	if err != nil {
		if errors.Is(err, errApprovalRuleNotFound) {
			d.SetId("")
			return nil
		}
		return err
	}
```

Reply:
I've added the global `errApprovalRuleNotFound`. Thanks!